File: Gemini/Intermediárias/validate_bst/validate_bst.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isValidBST(root):
    def helper(node, lower=float('-inf'), upper=float('inf')):
        if not node:
            return True
        val = node.val
        if val <= lower or val >= upper:
            return False
        if not helper(node.left, lower, val):
            return False
        return helper(node.right, val, upper)

    return helper(root)

def test_isValidBST():
    # Caso 1: Árvore válida
    root = TreeNode(10)
    root.left = TreeNode(5)
    root.right = TreeNode(15)
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(7)
    root.right.left = TreeNode(13)
    root.right.right = TreeNode(18)
    assert isValidBST(root) == True

    # Caso 2: Árvore inválida (valor duplicado)
    root = TreeNode(10)
    root.left = TreeNode(5)
    root.right = TreeNode(15)
    root.left.left = TreeNode(5)
    assert isValidBST(root) == False


if __name__ == "__main__":
    test_isValidBST()

File: ChatGPT/Intermediárias/validate_bst/validate_bst.cpp
#include <iostream>
#include <limits>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    // Função para verificar se uma árvore binária é uma BST
    bool isValidBST(TreeNode* root) {
        TreeNode* prev = nullptr;
        return inorder(root, prev);
    }

private:
    // Função auxiliar para fazer a travessia em ordem
    bool inorder(TreeNode* node, TreeNode*& prev) {
        if (node == nullptr) {
            return true;
        }

        // Visite a subárvore esquerda
        if (!inorder(node->left, prev)) {
            return false;
        }

        // Verifique se o nó atual é maior que o nó anterior na travessia em ordem
        if (prev != nullptr && node->val <= prev->val) {
            return false;
        }

        // Atualiza o nó anterior para o atual antes de visitar a subárvore direita
        prev = node;

        // Visite a subárvore direita
        return inorder(node->right, prev);
    }
};

int main() {
    // Criação de uma árvore binária que é uma BST
    TreeNode *root = new TreeNode(2);
    root->left = new TreeNode(1);
    root->right = new TreeNode(3);

    Solution sol;
    std::cout << "A árvore é uma BST válida: " << (sol.isValidBST(root) ? "Sim" : "Não") << std::endl;

    // Criação de uma árvore binária que não é uma BST
    TreeNode *root2 = new TreeNode(5);
    root2->left = new TreeNode(1);
    root2->right = new TreeNode(4);
    root2->right->left = new TreeNode(3);
    root2->right->right = new TreeNode(6);

    std::cout << "A segunda árvore é uma BST válida: " << (sol.isValidBST(root2) ? "Sim" : "Não") << std::endl;

    return 0;
}
